Initial version of generic migration guide (#1471)
chainguard-dev/edu: content/chainguard/migration-guides/migrating-to-chainguard-images.md

---
title: "Migrating Dockerfiles to Chainguard Images"
linktitle: "Migrating to Chainguard Images"
type: "article"
description: "Guidance on how to migrate existing Dockerfile workloads to use Chainguard Images"
date: 2024-03-25T15:56:52-07:00
lastmod: 2024-03-25T15:56:52-07:00
draft: false
tags: ["Images", "Product", "Conceptual"]
images: []

> menu:

Reviewer: ```suggestion
```

>   docs:

Reviewer: ```suggestion
```

>     parent: "concepts"

Reviewer: ```suggestion

```
I think this isn't necessary here?

weight: 100
toc: true
---

> ## Overview

Reviewer: ```suggestion

```
Your move, but I don't think this is necessary

> Based on the [Wolfi](/open-source/wolfi/overview/) Linux _undistro_, Chainguard Images have special features designed for increased security and provenance attestation. Depending on your current base image and custom commands, you may need some adjustments when migrating your current Dockerfile workloads to use Chainguard Images.

Reviewer: ```suggestion
Based on the [Wolfi](/open-source/wolfi/overview/) Linux _undistro_, Chainguard Images have special features designed for increased security and provenance attestation. Depending on your current base image and custom commands, you may need to make some adjustments when migrating your current Dockerfile workloads to use Chainguard Images.
```

A general migration process would involve the following steps:

> 1. **Identify the base image you need**. Look at the [reference documentation](/chainguard/chainguard-images/reference) to identify the image that is the closest match to what you currently use. You may also use [wolfi-base](https://edu.chainguard.dev/chainguard/chainguard-images/reference/wolfi-base) as a flexible starting point for your experimentation.

Reviewer: ```suggestion
1. **Identify the base image you need**. Check out the [reference documentation](/chainguard/chainguard-images/reference) to identify the image that is the closest match to what you currently use. You may also use [wolfi-base](/chainguard/chainguard-images/reference/wolfi-base) as a flexible starting point for your experimentation.
```

Author: good catch! thank you

2. **Try the `-dev` variant of the image first.** Chainguard Images typically have a **distroless** variant, which is very minimal and doesn't include `apk`, and a **dev** variant that contains tooling necessary to build applications and install new packages. Start with the **dev** variant or the **wolfi-base** image to have more room for customization.
3. **Identify packages you need to install**. Depending on your current base image, you may need to include additional packages to meet dependencies. Refer to the [Searching for Packages](#searching-for-packages) section for more details on how to find packages.
4. **Migrate to a distroless image**. Evaluate the option of using a Docker multi-stage build to create a final distroless image containing only what you need. Check the [Getting Started with Distroless images](/chainguard/chainguard-images/getting-started-distroless/) for more details of how to work with distroless images. Although not required, this process should give you a smaller image with additional safeguards.

> There are some differences in Wolfi's `busybox` and `coreutils` packages when compared to their counterparts in distros such as Debian or even Alpine. Some binaries and scripts are not included by default, which contributes to a smaller package size. Commands might still be available through separate packages, in order to keep images to a minimum.

Reviewer: ```suggestion
There are some differences in Wolfi's `busybox` and `coreutils` packages when compared to their counterparts in distros such as Debian or even Alpine. Some binaries and scripts are not included by default, which contributes to a smaller package size. This was done in order to keep images to a minimum, but be aware that some commands might still be available through separate packages.
```

The next sections of this page contain distro-specific information that should help you streamline the migration process from your current base images to Chainguard images.

## Migrating from Debian and Ubuntu Dockerfiles

> Chainguard Images are based on the [Alpine apk](https://wiki.alpinelinux.org/wiki/Package_management) ecosystem, which differs from Debian-based `apt` in several aspects. Some of these features contribute in making packages smaller and more accountable, resulting in smaller images with traceable provenance information based on cryptographic signatures. The page [Why apk](https://edu.chainguard.dev/open-source/wolfi/apk-package-manager/) from the official Wolfi documentation explains in more detail why we use apk.

Reviewer: ```suggestion
Chainguard Images are based on the [Alpine apk](https://wiki.alpinelinux.org/wiki/Package_management) ecosystem, which differs from Debian-based `apt` in several aspects. Some of these features contribute in making packages smaller and more accountable, resulting in smaller images with traceable provenance information based on cryptographic signatures. The page [Why apk](/open-source/wolfi/apk-package-manager/) from the official Wolfi documentation explains in more detail why we use apk.
```

If you are coming from a Debian-based Dockerfile, you'll need to adapt some of your commands to be compatible with the apk ecosystem:

| Command Description          | Debian-based Dockerfile | Wolfi-based Equivalent |
|------------------------------|-------------------------|------------------------|
| Install a package            | `apt install`           | `apk add`              |
| Remove a package             | `apt remove`            | `apk del`              |
| Update package manager cache | `apt update`            | `apk update`           |

> Our [Debian Compatibility](https://edu.chainguard.dev/chainguard/migration-guides/debian-compatibility/) page has a table listing common tools and their corresponding package(s) in both Wolfi and Debian distributions. For Ubuntu-based Dockerfiles, check our [Ubuntu Compatibility](https://edu.chainguard.dev/chainguard/migration-guides/ubuntu-compatibility/) page.

Reviewer: ```suggestion
Our [Debian Compatibility](/chainguard/migration-guides/debian-compatibility/) page has a table listing common tools and their corresponding package(s) in both Wolfi and Debian distributions. For Ubuntu-based Dockerfiles, check our [Ubuntu Compatibility](/chainguard/migration-guides/ubuntu-compatibility/) page.
```

## Migrating from Red Hat UBI Dockerfiles

> If you are coming from a Red Hat UBI (Universal Base Image) Dockerfile, you'll also need to adapt some of your commands to be compatible with the apk ecosystem. Wolfi uses BusyBox utilities, which offer a smaller footprint compared to GNU coreutils in Red Hat images. Our [Red Hat Compatibility](https://edu.chainguard.dev/chainguard/migration-guides/debian-compatibility/) page has a table listing common tools and their corresponding package(s) in both Wolfi and Red Hat distributions.

Reviewer: ```suggestion
If you are coming from a Red Hat UBI (Universal Base Image) Dockerfile, you'll also need to adapt some of your commands to be compatible with the apk ecosystem. Wolfi uses BusyBox utilities, which offer a smaller footprint compared to GNU coreutils in Red Hat images. Our [Red Hat Compatibility](/chainguard/migration-guides/debian-compatibility/) page has a table listing common tools and their corresponding package(s) in both Wolfi and Red Hat distributions.
```

If you are coming from a Red Hat UBI based Dockerfile, you'll need to adapt some of your commands to be compatible with the apk ecosystem:

| Command Description          | Red Hat UBI Dockerfile | Wolfi-based Equivalent |
|------------------------------|------------------------|------------------------|

> | Install a package            | `yum install`          | apk add                |

Reviewer: ```suggestion
| Install a package            | `yum install`          | `apk add`                |
```
for consistency

> | Remove a package             | `yum remove`           | apk del                |

Reviewer: ```suggestion
| Remove a package             | `yum remove`           | `apk del`                |
```

> | Update package manager cache | `yum makecache`        | apk update             |

Reviewer: ```suggestion
| Update package manager cache | `yum makecache`        | `apk update`             |
```

## Migrating from Alpine Dockerfiles

> If your Dockerfile is based on Alpine, the process for migrating to Chainguard Images should be more straightforward, since you're already using `apk` commands. Wolfi packages typically match what is available in Alpine, with some exceptions. For instance, the Wolfi busybox package is slimmer and doesn't include all tools available in Alpine's busybox. Check the [Alpine Compatibility](https://edu.chainguard.dev/chainguard/migration-guides/alpine-compatibility/) page for a list of common tools and their corresponding packages in Wolfi and Alpine.

Reviewer: Previously, "busybox" is styled as "BusyBox". From what I can tell it seems to be similar usage so the styling should be consistent throughout

Reviewer: ```suggestion
If your Dockerfile is based on Alpine, the process for migrating to Chainguard Images should be more straightforward, since you're already using `apk` commands. Wolfi packages typically match what is available in Alpine, with some exceptions. For instance, the Wolfi busybox package is slimmer and doesn't include all tools available in Alpine's busybox. Check the [Alpine Compatibility](/chainguard/migration-guides/alpine-compatibility/) page for a list of common tools and their corresponding packages in Wolfi and Alpine.
```

## Searching for Packages
Packages from Debian and other base distributions might have a different name in Wolfi. To search for packages, log into an ephemeral container based on `cgr.dev/chainguard/wolfi-base`:

```shell
docker run -it --rm --entrypoint /bin/sh cgr.dev/chainguard/wolfi-base
```

Then, run `apk update` to update the local apk cache with latest Wolfi packages:

```shell

> # apk update

Reviewer: ```suggestion
apk update
```
These commands all work but it makes sense to me to break up these commands from their expected outputs

Reviewer 2: `#` is root's prompt, so this shows nearly exactly what you'd expect to see. A lot of people do this with `#` or`$` to indicate running as root or as a user. I wish they wouldn't.

This is a small bugbear of mine, when you view the rendered file there is a "copy" shortcut button on the command's text box and it copies _everything_ . Including the `#` and command response. Which makes the copy button useless, because you can't paste commands prefixed with # and pasting the result of someone else's command into your terminal just makes a mess.

Further down where we `apk search` in one commands block and the results are in the next is better IMHO (or would be if the `apk search` didn't have a `#` !) Because you can copy _just_ the command, without the response.

Rather than repeating the command from the preamble in the example, my suggestion would be something like:

> 
> Then, run:
> ```
> apk update
> ```
> to update the local cache, If the update is successful, you'll see a response like:
> ```
> fetch https:... 
> ......
> OK 1234 distinct packages available
> ```

For ultimate "perfection" could we remove the "copy" button from the response text blocks? If someone wants to copy the response, they can select/copy manually. (This is strictly in the "nice to have" category if it is hard).

fetch https://packages.wolfi.dev/os/x86_64/APKINDEX.tar.gz
[https://packages.wolfi.dev/os]
OK: 46985 distinct packages available
```
Now you can use `apk search` to look for packages. The following example searches for PHP 8.2 XML extensions:

```shell

> # apk search php*8.2*xml*

Reviewer: ```suggestion
apk search php*8.2*xml*
```
Here especially since it's already separate from its output, I don' tthink the #'s are necessary

```
You should get output similar to this:

> ```shell

Reviewer: shell is unnecessary here

php-8.2-simplexml-8.2.17-r0
php-8.2-simplexml-config-8.2.17-r0
php-8.2-xml-8.2.17-r0
php-8.2-xml-config-8.2.17-r0
php-8.2-xmlreader-8.2.17-r0
php-8.2-xmlreader-config-8.2.17-r0
php-8.2-xmlwriter-8.2.17-r0
php-8.2-xmlwriter-config-8.2.17-r0
php-simplexml-8.2.11-r1
php-xml-8.2.11-r1
php-xmlreader-8.2.11-r1
php-xmlwriter-8.2.11-r1
```

### Searching which package has a command
To search in which package you can find a command, you can use the syntax `apk search cmd:command-name`. For instance, if you want to discover which package has the command `useradd`, you can use:

```shell

> # apk search cmd:useradd

Reviewer: ```suggestion
apk search cmd:useradd
```

```
You'll get output indicating that the `shadow` package has the command you are looking for.

> ```shell

Reviewer: ```suggestion
```
```

Reviewer: I don't think it's showing up correclty, but I don't think the `shell` is necessary ehre

shadow-4.15.1-r0
```

### Searching for package dependencies
To check for package dependencies, you can use the syntax `apk search -R info package`. For example, to search which packages are listed as dependencies for the `shadow` package that we've seen in the previous section, you can run:

```shell
apk -R info shadow
```
And this will give you a list of dependencies for each version of the `shadow` package currently available.

### Searching for packages that include a shared object
To search which packages include a shared object, you can use the syntax `apk search so:shared-library`. As an example, if you want to check which packages include the `libxml2` shared library, you can run something like:

```shell
apk search so:libxml2.so*
```
And this should give you output indicating that this shared object is included within the `libxml2-2.12.6-r0` package.

For detailed information about apk options and flags when searching for packages, check the [official documentation](https://docs.alpinelinux.org/user-handbook/0.1a/Working/apk.html#_searching_for_packages).

## Resources to Learn More

> Our [Getting Started Guides](https://edu.chainguard.dev/chainguard/chainguard-images/getting-started/) have detailed examples for different language ecosystems and stacks. Make sure to also check image-specific documentation in our [reference docs](https://edu.chainguard.dev/chainguard/chainguard-images/reference/).

Reviewer: ```suggestion
Our [Getting Started Guides](/chainguard/chainguard-images/getting-started/) have detailed examples for different language ecosystems and stacks. Make sure to also check image-specific documentation in our [reference docs](/chainguard/chainguard-images/reference/).
```

> If you can't find an image that is a good match for your use case, or if your build has dependencies that cannot be met with the regular catalog, [get in touch with us](https://www.chainguard.dev/contact) for alternative options.

Reviewer: ```suggestion
If you can't find an image that is a good match for your use case, or if your build has dependencies that cannot be met with the regular catalog, [get in touch with us](https://www.chainguard.dev/contact?utm_source=docs) for alternative options.
```
